ydb/core/blobstorage/vdisk/balance/balancing_actor.cpp

```diff
@@ -1,29 +1,220 @@
+
+            SenderId = ctx.Register(CreateSenderActor(SelfId(), std::move(SendOnMainParts), QueueActorMapPtr, Ctx));
+            ActiveActors.Insert(SenderId, __FILE__, __LINE__, ctx, NKikimrServices::BLOBSTORAGE);
+
+            DeleterId = ctx.Register(CreateDeleterActor(SelfId(), std::move(TryDeleteParts), QueueActorMapPtr, Ctx));
+            ActiveActors.Insert(DeleterId, __FILE__, __LINE__, ctx, NKikimrServices::BLOBSTORAGE);
+
             Become(&TThis::StateFunc);
+            Schedule(TDuration::Seconds(1), new NActors::TEvents::TEvWakeup());
         }
 
+        void StartBalancing(const TActorContext &ctx) {
+            LOG_DEBUG_S(ctx, NKikimrServices::BS_VDISK_BALANCING,
+                        Ctx->VCtx->VDiskLogPrefix << "Ask repl token");
+            if (!Send(MakeBlobStorageReplBrokerID(), new TEvQueryReplToken(Ctx->VDiskCfg->BaseInfo.PDiskId))) {
+                HandleReplToken(ctx);
+            }
+        }
+
+        void HandleReplToken(const TActorContext &ctx) {
+            LOG_DEBUG_S(ctx, NKikimrServices::BS_VDISK_BALANCING,
+                        Ctx->VCtx->VDiskLogPrefix << "Repl token acquired");
+            DoJobQuant(ctx);
+            Send(MakeBlobStorageReplBrokerID(), new TEvReleaseReplToken);
+            Schedule(TDuration::Seconds(1), new NActors::TEvents::TEvWakeup());
+        }
+
+        void Handle(NActors::TEvents::TEvCompleted::TPtr ev, const TActorContext &ctx) {
+            switch (ev->Get()->Id) {
+                case SENDER_ID: {
+                    Stats.SendCompleted = true;
+                    break;
+                }
+                case DELETER_ID: {
+                    Stats.DeleteCompleted = true;
+                    break;
+                }
+                default:
+                    Y_ABORT("Unexpected id");
+            }
+            if (Stats.SendCompleted && Stats.DeleteCompleted) {
+                LOG_DEBUG_S(ctx, NKikimrServices::BS_VDISK_BALANCING,
+                            Ctx->VCtx->VDiskLogPrefix << "Balancing completed");
+                Send(Ctx->SkeletonId, new TEvStartBalancing());
+                Send(SelfId(), new NActors::TEvents::TEvPoison);
+            }
+        }
+
+        void Handle(TEvProxyQueueState::TPtr ev) {
+            const TVDiskID& vdiskId = ev->Get()->VDiskId;
+            if (ev->Get()->IsConnected) {
+                ConnectedVDisks.insert(vdiskId);
+            } else {
+                ConnectedVDisks.erase(vdiskId);
+            }
+        }
+
+        void Handle(TEvVGenerationChange::TPtr ev) {
+            Cerr << "TEvVGenerationChange" << Endl;
+            // forward message to queue actors
+            TEvVGenerationChange *msg = ev->Get();
+            for (const auto& kv : *QueueActorMapPtr) {
+                Send(kv.second, msg->Clone());
+            }
+            Ctx->GInfo = msg->NewInfo;
```

Comment:
Storing something in Ctx this way is bad idea. It is assumed to be thread-safe. Either it is read-only, or it is stored locally. So it's better to keep group info as local variable.

Reply:
It is a copypaste from replication actor https://github.com/ydb-platform/ydb/blob/2380e383a99584b10ed0ac5ac570fd26a6593a62/ydb/core/blobstorage/vdisk/repl/blobstorage_repl.cpp#L545
Anyway it is not very clear what is a good way to update GInfo, because it is using in sender and deleter actors too